tokio/src/runtime/task/core.rs

```diff
@@ -95,31 +95,30 @@ impl<T: Future, S: Schedule> Cell<T, S> {
 }
 
 impl<T: Future, S: Schedule> Core<T, S> {
-    /// If needed, bind a scheduler to the task.
+    /// Bind a scheduler to the task.
     ///
-    /// This only happens on the first poll.
-    pub(super) fn bind_scheduler(&self, task: Task<S>) {
-        use std::mem::ManuallyDrop;
-
-        // TODO: it would be nice to not have to wrap with a ManuallyDrop
-        let task = ManuallyDrop::new(task);
-
+    /// This only happens on the first poll and must be preceeded by a call to
+    /// `is_bound` to determine if binding is appropriate or not.
+    ///
+    /// # Safety
+    ///
+    /// Binding must not be done concurrently since it will mutate the task
+    /// core.
+    pub(super) unsafe fn bind_scheduler(&self, task: Task<S>) {
```

Comment:
I think some of the motivation for not marking the function as unsafe may have been that this would also make the whole function body implicitly `unsafe`, whereas when it is safe to call, the raw pointer assignment must still be explicitly marked as unsafe.

Reply:
virtually the entire task module's internal functions require the caller to uphold some safety invariant. Because of this, we opted to omit `unsafe` from module private fns. The reasoning being it was better to identify where invariants needed to be upheld to fns **external** to the module. The safety boundary should be the module and not the individual function.

That said, I skimmed the rest of the change and it makes sense at a glance.

Reply:
Changed back.